Go 1.6 compatibility

https://tip.golang.org/doc/go1.6#cgo
```
$ go version
go version go1.6rc2 darwin/amd64
$ go test
--- FAIL: TestVersionInfo (0.00s)
	core_test.go:20: protocol should be in [dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smtp smtps telnet tftp scp sftp] and is smb.
	core_test.go:20: protocol should be in [dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smtp smtps telnet tftp scp sftp] and is smbs.

--- FAIL: TestCallbackFunction (0.00s)
panic: runtime error: cgo argument has Go pointer to Go pointer [recovered]
	panic: runtime error: cgo argument has Go pointer to Go pointer

goroutine 19 [running]:
testing.tRunner.func1(0xc8200fa000)
	/usr/local/go/src/testing/testing.go:467 +0x192
github.com/andelf/go-curl._cgoCheckPointer0(0x42ece00, 0xc8200da0e0, 0x0, 0x0, 0x0, 0x6000000)
	??:0 +0x4d
github.com/andelf/go-curl.(*CURL).Setopt(0xc8200da0e0, 0x4e2b, 0x4343140, 0xc8200d01e0, 0x0, 0x0)
	/Users/ottob/go/src/github.com/andelf/go-curl/easy.go:197 +0x10e7
github.com/andelf/go-curl.TestCallbackFunction(0xc8200fa000)
	/Users/ottob/go/src/github.com/andelf/go-curl/easy_test.go:45 +0x1a7
testing.tRunner(0xc8200fa000, 0x45ca0b0)
	/usr/local/go/src/testing/testing.go:473 +0x98
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:582 +0x892
exit status 2
FAIL	github.com/andelf/go-curl	0.019s
```

--- a/c-callback.c
+++ b/c-callback.c
@@ -5,13 +5,7 @@
 
 /* for OPT_HEADERFUNCTION */
 size_t header_function( char *ptr, size_t size, size_t nmemb, void *ctx) {
-    void *go_header_func = (void *)goGetCurlField((GoUintptr)ctx, "headerFunction");
-    GoInterface *userdata = (GoInterface *)goGetCurlField((GoUintptr)ctx, "headerData");
-
-    if (userdata == NULL) {
-	return goCallWriteFunctionCallback(go_header_func, ptr, size*nmemb, goNilInterface());
-    }
-    return goCallWriteFunctionCallback(go_header_func, ptr, size*nmemb, *userdata);
+	return goCallHeaderFunction(ptr, size*nmemb, ctx);
 }
 
 void *return_header_function() {
@@ -21,13 +15,7 @@ void *return_header_function() {
 
 /* for OPT_WRITEFUNCTION */
 size_t write_function( char *ptr, size_t size, size_t nmemb, void *ctx) {
-    void *go_write_func = (void *)goGetCurlField((GoUintptr)ctx, "writeFunction");
-    GoInterface *userdata = (GoInterface *)goGetCurlField((GoUintptr)ctx, "writeData");
-
-    if (userdata == NULL) {
-	return goCallWriteFunctionCallback(go_write_func, ptr, size*nmemb, goNilInterface());
-    }
-    return goCallWriteFunctionCallback(go_write_func, ptr, size*nmemb, *userdata);
+	return goCallWriteFunction(ptr, size*nmemb, ctx);
 }
 
 void *return_write_function() {
@@ -36,13 +24,7 @@ void *return_write_function() {
 
 /* for OPT_READFUNCTION */
 size_t read_function( char *ptr, size_t size, size_t nmemb, void *ctx) {
-    void *go_read_func = (void *)goGetCurlField((GoUintptr)ctx, "readFunction");
-    GoInterface *userdata = (GoInterface *)goGetCurlField((GoUintptr)ctx, "readData");
-
-    if (userdata == NULL) {
-	return goCallReadFunctionCallback(go_read_func, ptr, size*nmemb, goNilInterface());
-    }
-    return goCallReadFunctionCallback(go_read_func, ptr, size*nmemb, *userdata);
+	return goCallReadFunction(ptr, size*nmemb, ctx);
 }
 
 void *return_read_function() {
@@ -52,15 +34,7 @@ void *return_read_function() {
 
 /* for OPT_PROGRESSFUNCTION */
 int progress_function(void *ctx, double dltotal, double dlnow, double ultotal, double ulnow) {
-    void *go_progress_func = (void *)goGetCurlField((GoUintptr)ctx, "progressFunction");
-    GoInterface *clientp = (GoInterface *)goGetCurlField((GoUintptr)ctx, "progressData");
-
-    if (clientp == NULL) {
-	return goCallProgressCallback(go_progress_func, goNilInterface(),
-				    dltotal, dlnow, ultotal, ulnow);
-    }
-    return goCallProgressCallback(go_progress_func, *clientp,
-				dltotal, dlnow, ultotal, ulnow);
+	return goCallProgressFunction(dltotal, dlnow, ultotal, ulnow, ctx);
 }
 
 void *return_progress_function() {

--- a/callback.go
+++ b/callback.go
@@ -9,78 +9,45 @@ package curl
 import "C"
 
 import (
-	"reflect"
 	"unsafe"
 )
 
-//export goGetCurlField
-func goGetCurlField(p uintptr, cname *C.char) uintptr {
-	name := C.GoString(cname)
-	curl := (*CURL)(unsafe.Pointer(p))
-	switch name {
-	case "readFunction":
-		return reflect.ValueOf(curl.readFunction).Pointer()
-	case "headerFunction":
-		return reflect.ValueOf(curl.headerFunction).Pointer()
-	case "writeFunction":
-		return reflect.ValueOf(curl.writeFunction).Pointer()
-	case "progressFunction":
-		return reflect.ValueOf(curl.progressFunction).Pointer()
-	case "headerData":
-		return uintptr(unsafe.Pointer(curl.headerData))
-	case "writeData":
-		return uintptr(unsafe.Pointer(curl.writeData))
-	case "readData":
-		return uintptr(unsafe.Pointer(curl.readData))
-	case "progressData":
-		return uintptr(unsafe.Pointer(curl.progressData))
-	}
-
-	warnf("Field not found: %s", name)
-	return 0
-}
-
-//export goNilInterface
-func goNilInterface() interface{} {
-	return nil
-}
-
-// callback functions
-//export goCallWriteFunctionCallback
-func goCallWriteFunctionCallback(f *func([]byte, interface{}) bool,
-	ptr *C.char,
-	size C.size_t,
-	userdata interface{}) uintptr {
+//export goCallHeaderFunction
+func goCallHeaderFunction(ptr *C.char, size C.size_t, ctx unsafe.Pointer) uintptr {
+	curl := (*CURL)(ctx)
 	buf := C.GoBytes(unsafe.Pointer(ptr), C.int(size))
-	ok := (*f)(buf, userdata)
-	if ok {
+	if (*curl.headerFunction)(buf, curl.headerData) {
 		return uintptr(size)
 	}
-	//return uintptr(C.CURL_MAX_WRITE_SIZE + 1)
 	return C.CURL_WRITEFUNC_PAUSE
 }
 
-//export goCallProgressCallback
-func goCallProgressCallback(f *func(float64, float64, float64, float64, interface{}) bool,
-	userdata interface{},
-	dltotal, dlnow, ultotal, ulnow C.double) int {
-	// fdltotal, fdlnow, fultotal, fulnow
-	ok := (*f)(float64(dltotal), float64(dlnow), float64(ultotal), float64(ulnow), userdata)
-	// non-zero va lue will cause libcurl to abort the transfer and return Error
-	if ok {
+//export goCallWriteFunction
+func goCallWriteFunction(ptr *C.char, size C.size_t, ctx unsafe.Pointer) uintptr {
+	curl := context_map[uintptr(ctx)]
+	buf := C.GoBytes(unsafe.Pointer(ptr), C.int(size))
+	if (*curl.writeFunction)(buf, curl.writeData) {
+		return uintptr(size)
+	}
+	return C.CURL_WRITEFUNC_PAUSE
+}
+
+//export goCallProgressFunction
+func goCallProgressFunction(dltotal, dlnow, ultotal, ulnow C.double, ctx unsafe.Pointer) int {
+	curl := context_map[uintptr(ctx)]
+	if (*curl.progressFunction)(float64(dltotal), float64(dlnow),
+		                 float64(ultotal), float64(ulnow),
+		                 curl.progressData) {
 		return 0
 	}
 	return 1
 }
 
-//export goCallReadFunctionCallback
-func goCallReadFunctionCallback(f *func([]byte, interface{}) int,
-	ptr *C.char,
-	size C.size_t,
-	userdata interface{}) uintptr {
-	// TODO code cleanup
+//export goCallReadFunction
+func goCallReadFunction(ptr *C.char, size C.size_t, ctx unsafe.Pointer) uintptr {
+	curl := context_map[uintptr(ctx)]
 	buf := C.GoBytes(unsafe.Pointer(ptr), C.int(size))
-	ret := (*f)(buf, userdata)
+	ret := (*curl.readFunction)(buf, curl.readData)
 	str := C.CString(string(buf))
 	defer C.free(unsafe.Pointer(str))
 	if C.memcpy(unsafe.Pointer(ptr), unsafe.Pointer(str), C.size_t(ret)) == nil {

--- a/easy.go
+++ b/easy.go
@@ -66,7 +66,6 @@ import (
 	"fmt"
 	"mime"
 	"path"
-	"reflect"
 	"unsafe"
 )
 
@@ -105,21 +104,27 @@ type CURL struct {
 	progressFunction              *func(float64, float64, float64, float64, interface{}) bool
 	fnmatchFunction               *func(string, string, interface{}) int
 	// callback datas
-	headerData, writeData, readData, progressData, fnmatchData *interface{}
+	headerData, writeData, readData, progressData, fnmatchData interface{}
 	// list of C allocs
 	mallocAllocs []*C.char
 }
 
+var context_map = make(map[uintptr]*CURL)
+
 // curl_easy_init - Start a libcurl easy session
 func EasyInit() *CURL {
 	p := C.curl_easy_init()
-	return &CURL{handle: p, mallocAllocs: make([]*C.char, 0)} // other field defaults to nil
+	c := &CURL{handle: p, mallocAllocs: make([]*C.char, 0)} // other field defaults to nil
+	context_map[uintptr(p)] = c
+	return c
 }
 
 // curl_easy_duphandle - Clone a libcurl session handle
 func (curl *CURL) Duphandle() *CURL {
-	p := curl.handle
-	return &CURL{handle: C.curl_easy_duphandle(p)}
+	p := C.curl_easy_duphandle(curl.handle)
+	c := &CURL{handle: p}
+	context_map[uintptr(p)] = c
+	return c
 }
 
 // curl_easy_cleanup - End a libcurl easy session
@@ -127,6 +132,7 @@ func (curl *CURL) Cleanup() {
 	p := curl.handle
 	C.curl_easy_cleanup(p)
 	curl.MallocFreeAfter(0)
+	delete(context_map, uintptr(p))
 }
 
 // curl_easy_setopt - set options for a curl easy handle
@@ -140,16 +146,16 @@ func (curl *CURL) Setopt(opt int, param interface{}) error {
 	switch {
 	// not really set
 	case opt == OPT_READDATA: // OPT_INFILE
-		curl.readData = &param
+		curl.readData = param
 		return nil
 	case opt == OPT_PROGRESSDATA:
-		curl.progressData = &param
+		curl.progressData = param
 		return nil
 	case opt == OPT_HEADERDATA: // also known as OPT_WRITEHEADER
-		curl.headerData = &param
+		curl.headerData = param
 		return nil
 	case opt == OPT_WRITEDATA: // OPT_FILE
-		curl.writeData = &param
+		curl.writeData = param
 		return nil
 
 	case opt == OPT_READFUNCTION:
@@ -158,8 +164,7 @@ func (curl *CURL) Setopt(opt int, param interface{}) error {
 
 		ptr := C.return_read_function()
 		if err := newCurlError(C.curl_easy_setopt_pointer(p, C.CURLoption(opt), ptr)); err == nil {
-			return newCurlError(C.curl_easy_setopt_pointer(p, OPT_READDATA,
-				unsafe.Pointer(reflect.ValueOf(curl).Pointer())))
+			return newCurlError(C.curl_easy_setopt_pointer(p, OPT_READDATA, unsafe.Pointer(curl.handle)))
 		} else {
 			return err
 		}
@@ -170,8 +175,7 @@ func (curl *CURL) Setopt(opt int, param interface{}) error {
 
 		ptr := C.return_progress_function()
 		if err := newCurlError(C.curl_easy_setopt_pointer(p, C.CURLoption(opt), ptr)); err == nil {
-			return newCurlError(C.curl_easy_setopt_pointer(p, OPT_PROGRESSDATA,
-				unsafe.Pointer(reflect.ValueOf(curl).Pointer())))
+			return newCurlError(C.curl_easy_setopt_pointer(p, OPT_PROGRESSDATA, unsafe.Pointer(curl.handle)))
 		} else {
 			return err
 		}
@@ -182,8 +186,7 @@ func (curl *CURL) Setopt(opt int, param interface{}) error {
 
 		ptr := C.return_header_function()
 		if err := newCurlError(C.curl_easy_setopt_pointer(p, C.CURLoption(opt), ptr)); err == nil {
-			return newCurlError(C.curl_easy_setopt_pointer(p, OPT_HEADERDATA,
-				unsafe.Pointer(reflect.ValueOf(curl).Pointer())))
+			return newCurlError(C.curl_easy_setopt_pointer(p, OPT_HEADERDATA, unsafe.Pointer(curl.handle)))
 		} else {
 			return err
 		}
@@ -194,8 +197,7 @@ func (curl *CURL) Setopt(opt int, param interface{}) error {
 
 		ptr := C.return_write_function()
 		if err := newCurlError(C.curl_easy_setopt_pointer(p, C.CURLoption(opt), ptr)); err == nil {
-			return newCurlError(C.curl_easy_setopt_pointer(p, OPT_WRITEDATA,
-				unsafe.Pointer(reflect.ValueOf(curl).Pointer())))
+			return newCurlError(C.curl_easy_setopt_pointer(p, OPT_WRITEDATA, unsafe.Pointer(curl.handle)))
 		} else {
 			return err
 		}

